Rigid Origami Folding

asahidari opened the pull request:
## Addressed problem description

This PR enable us to simulate wide range of rigid origami fold. Previously, I added #3375 for only one type of folding, but this PR can offer various type of origami folding. Also, this is related to #1381 'WISHLIST' of origami fold.

## Solution description

I added a script node to SNLite_templates/utils directory to avoid any influence or changes to other modules.
Inputs and outputs of this script node are following. I see this link as a reference: https://www.grasshopper3d.com/forum/topics/new-origami-example?groupUrl=kangaroo&groupId=2985220%3AGroup%3A120977&id=2985220%3ATopic%3A973616&page=3

- [in] obj_in (object)
- [in] valleys (edge list)
- [in] valley angle (scalar value)
- [in] mountains (edge list)
- [in] mountain angle (scalar value)
- [in] folding (scalar value, ratio of folding)
- [in] step (scalar value to use as delta angle in integral calculation )
- [in] fixed_face (index of a mesh polygon. Face rotation begin at this face)
- [out] verts (moved vertices)
- [out] edges (edges)
- [out] faces (faces)

![rigid_origami_folding_screen_shot](https://user-images.githubusercontent.com/64673405/88299105-f297fc00-cd3c-11ea-920a-b87c5f8b874e.png)

Next is an example of the node tree.
![rigid_origami_folding_node_tree_screen_shot](https://user-images.githubusercontent.com/64673405/88301513-aac6a400-cd3f-11ea-9d2d-8ed3826dbcee.png)


In this script, these classes are included.
- ObjectParams (Object wrapper)
- CreaseLines (Manager of crease lines)
- InsideVertex (Instance of each vertices inside of the target paper)
- FoldAngleCalculator (Calculator of each edge's fold angles)
- FaceRotation (Handler to operate face rotation)

Limitation:
- Some types of folding may not be done, which need some extra strength.
For example, https://www.youtube.com/watch?v=IP7WbV_yLgQ or https://www.foldfactory.com/offering/twist .
- It may be difficult to find where error occurred. For example, when using wrong edge list or opposite valleys/mountains.
- It's not always right shape because this calculation include differentiation and integration, but these are semi-pseudo implementation using only NumPy library.

## Preflight checklist

Put an x letter in each brackets when you're done this item:

- [x] Code changes complete.
- [] Code documentation complete.
- [x] Documentation for users complete (or not required, if user never sees these changes).
- [x ] Manual testing done. 
- [ ] Unit-tests implemented.
- [x] Ready for merge.

asahidari committed:
add rigid_origami_folding.py script to SNLite_templates

asahidari commented:
These are some examples.

1. Twisted square folding.
![rigid_origami_fold_example1](https://user-images.githubusercontent.com/64673405/88309962-11e95600-cd4a-11ea-8715-d8a794a49705.png)

2. Water bomb tessellation no.1
![rigid_origami_fold_example2](https://user-images.githubusercontent.com/64673405/88310026-29284380-cd4a-11ea-8f8b-dc111bd5f467.png)

3. Water bomb tessellation no.2
![rigid_origami_fold_example3](https://user-images.githubusercontent.com/64673405/88310045-2fb6bb00-cd4a-11ea-8e82-3e01d23ed849.png)

vicdoval commented:
@asahidari  This is totally awesome!!! 
I wont be able to check the code in a couple of weeks in case anyone else want to check it before merging but I love the features this PR brings.

Also very interesting of the grasshopper example you provide is the mix with the kangaroo plugin that could be extrapolated to the Pulga Physics node.

Another idea @asahidari would be creating a full node instead of a SNL script :)  (maybe i will do it if you don't want to)

asahidari commented:
I'm glad to hear that, and it's no problem to take some weeks. This has nearly 500 lines and may have some redundant codes.

Pulga Physics is fascinating, but cannot afford to understand it quickly when I writing this. If the origami node can collaborate the physics node, it is interesting and the tree node may become more simple. 

Currently, I think I do not have enough understanding sverchok's structure. So if I can easily write a full node and find where to place it, I want to do that. But if the collaboration with the physics node is not easy and there are some worries to collapse existing codes, please consider you would change that.

-----------
note about extra parameter:
step: I added this, because some folding process are different when this value changed. But it can be fixed value, like pi*0.05.
fixed face: Other faces are rotated following edge angles, but this face will not move as a start point of rotation. This also can be fixed, for example, at the center of the paper.

enzyme69 commented:
Maybe I just hit that merge button and test it to make crane origami... ☝️

asahidari commented:
This node is currently deployed in nodes_scripts/SNLite_templates/utils directory. So you can test it with copying whole text of 'rigid_origami_folding.py' and paste it on Blender's text editor to use in SNLite script node.

But unfortunately, crane folding seems not to work well :( .
![crane_strange_folding](https://user-images.githubusercontent.com/64673405/88449204-ef595900-ce7f-11ea-9056-ca6c8848752a.png)

It seems this script still has some bugs or limitations  (probably in face rotation process), because the motion is changed when the parameters 'step' and 'fixed_face' changed. I'll check it, but it might have good result when using with the physics node.

asahidari commented:
> But unfortunately, crane folding seems not to work well :( .
> ...
> .... I'll check it, but it might have good result when using with the physics node.

As I checked my code and read the original article again, Crane folding in this way seems not to be successful. The author of the [original paper](https://origami.c.u-tokyo.ac.jp/~tachi/cg/SimulationOfRigidOrigami_tachi_4OSME.pdf) says in chapter.4:

`Many models can be folded to the desired state in this system, while many cannot be folded. ... For example, Miura fold, waterbomb tessellation, and pleated hypars are smoothly folded in the simulation, while a crane with finishing, four-crane base or tatou cannot be not folded to the final state.`

When you search for crane folding on line, [Origami Simulator](http://www.amandaghassaei.com/projects/origami_simulator/) can fold crane, but the app uses another method (using another type of crease line and pin-jointed truss model) and GPU to do more complicated calculation. And I could not find crane fold process using Rhino+Grasshopper, Houdini. So I think it is currently still tough to realize it.

enzyme69 commented:
<img width="966" alt="Screen Shot 2020-07-27 at 11 37 53 am" src="https://user-images.githubusercontent.com/3952687/88495557-a0d1c900-cffd-11ea-9b35-d31d36a8aae2.png">

I tested the node, very often when folding algorithm failled, it does not give error and no result.
https://youtu.be/X614Jv1kpX8

asahidari commented:
> I tested the node, very often when folding algorithm failed, it does not give error and no result.
> ...
> I want to try a simple paper airplane, but somewhat so far cannot get result

I'm sorry that the node has some non-user-friendly interface. It's actually not easy to know where is wrong.

But it seems better to begin with simple fold patterns. Selecting crease lines (valley/mountain lines) randomly may confuse you and sometimes it actually cannot be foldable, because the angles of the edges around vertices inside strictly influences each other.

Searching for images with words like "crease line origami" may show you some fold patterns. As you know, crane and some patterns cannot be foldable now, but others (like Water bomb tessellations, Miura-fold, Yoshimura-fold and etc.) may show you the good folding process.

nortikin commented:
![Снимок экрана от 2020-07-27 22-29-12](https://user-images.githubusercontent.com/5783432/88583240-a3154100-d058-11ea-8bc0-1592155271c0.png)
```python
out_outer = []
out_inner = []
for i,v in enumerate(V1[0].data.vertices[:]):
    if 0 in [g.group for g in v.groups.values()]:
        out_outer.append(i)
    if 1 in [g.group for g in v.groups.values()]:
        out_inner.append(i)
append([out_outer,out_inner])
```
```python
out_outer = []
out_inner = []
for k,i in enumerate(V1[0]):
    if i[0] in V2[0][0] and i[1] in V2[0][0]:
        out_outer.append(k)
    if i[0] in V2[0][1] and i[1] in V2[0][1]:
        out_inner.append(k)
append([out_outer,out_inner])
```

nortikin commented:
it is easily to rewrite in SNL and use without exec node

nortikin commented:
![Снимок экрана от 2020-07-27 23-07-15](https://user-images.githubusercontent.com/5783432/88587025-f473ff00-d05d-11ea-9bd5-416719561009.png)

nortikin commented:
```python
outer = []
inner = []
for i,v in enumerate(object.data.vertices[:]):
    if 0 in [g.group for g in v.groups.values()]:
        outer.append(i)
    if 1 in [g.group for g in v.groups.values()]:
        inner.append(i)

mount = []
valley = []
for k,i in enumerate(object.data.edges[:]):
    if i.vertices[0] in outer and i.vertices[1] in outer:
        mount.append(i)
    if i.vertices[0] in inner and i.vertices[1] in inner:
        valley.append(i)
```

nortikin commented on node_scripts/SNLite_templates/utils/rigid_origami_folding.py:
```diff
@@ -0,0 +1,486 @@
+
+#################################
+
+# create tuple list to avoid type mismatch
```
here we can grub objects vertex groups

enzyme69 commented:
@nortikin Can you supply the blend of the paperplane? Or make video about it?

enzyme69 commented:
@nortikin Can you try?
https://parametrichouse.com/origami-world-2/

What's the sane way to do that?

I tried the SNL Rigid Origami on a Cylinder (not plane) and Blender frozen.

enzyme69 commented:
[paperplane_folding_study_002.blend.zip](https://github.com/nortikin/sverchok/files/4985656/paperplane_folding_study_002.blend.zip)


Not working @nortikin 
<img width="1212" alt="Screen Shot 2020-07-28 at 12 42 42 pm" src="https://user-images.githubusercontent.com/3952687/88613008-e1017c00-d0cf-11ea-9545-8c7a6166958c.png">

enzyme69 commented:
Aha....... 😤 
<img width="982" alt="Screen Shot 2020-07-28 at 12 59 48 pm" src="https://user-images.githubusercontent.com/3952687/88614049-3a6aaa80-d0d2-11ea-9373-3bb2d217f965.png">

enzyme69 commented:
<img width="933" alt="Screen Shot 2020-07-28 at 1 23 32 pm" src="https://user-images.githubusercontent.com/3952687/88615562-8cf99600-d0d5-11ea-9fbd-bb5950b1a8bf.png">

asahidari commented:
> solved with vertex groups
> two groups, you can decide what is valley and hill.
> ...

Thank you for the suggestions. It seems good idea. But I'm afraid some 'peripheral' edges might be included.
(In this example, edge (2, 3) is not crease line, but included.)
![Screen Shot 2020-07-29 at 1 18 01](https://user-images.githubusercontent.com/64673405/88692777-6b31fa80-d139-11ea-897f-01bc793e5c93.png)

> Without mount/valley sockets. If we working with objects. But if we working with Sverchok mesh - it is another story.

Currently, this script is a SNLite script and that is why I put an 'object' to in-socket. But if this can be a full node and used with other nodes, it seems useful to use mesh instead of object.

asahidari commented:
I just posted some folding examples in my blog. I hope these examples would be helpful to use this node.

http://asahidari.hatenablog.com/entry/2020/07/29/010410?_ga=2.66962478.1020466821.1595564630-1234164817.1595564630

enzyme69 commented:
Great blog post showing some of the interesting paper folding and its patterns!! 

Paper Crane is actually possible?

nortikin commented:
@asahidari 
1.Did not found edges groups in blender 👎 (((
2.Agree, edge can be red as grouped. It could also included in both groups \o_O/
3.It works for me.
4.Case of inner sverchok data. We can insert some GP tricks with template, by the way.
5.Maybe to leave both approaches? object and sverchok.
6.No need to enter edges themselves, we can insert edges' indices. not to filter, less steps for user.
7.Thanks

asahidari commented:
> @asahidari
> 1.Did not found edges groups in blender 👎 (((
> 2.Agree, edge can be red as grouped. It could also included in both groups \o_O/
> 3.It works for me.

This node simply ignores edges when they have a face only one side of them. (it assumes that edges inside of the paper must have faces both side of them.) But when error occurs, existence of these edges may be confusing or misleading.

> 4.Case of inner sverchok data. We can insert some GP tricks with template, by the way.

GP(Grease Pencil ?) seems to have many possibilities, but I cannot quickly get ideas how to use it because I'm not used to it :-(.

> 5.Maybe to leave both approaches? object and sverchok.
> 6.No need to enter edges themselves, we can insert edges' indices. not to filter, less steps for user.

Yes, the node does not need object socket now. Instead, vertices, edges and faces are needed to calculate angles and rotating faces.

Another idea:
Blender has [edge crease](https://docs.blender.org/manual/en/2.79/modeling/meshes/editing/edges.html#edge-crease) and the edge can have values between -1 to 1. When users set the crease values to all edges inside of the paper and the node uses the values (multiplying with math.pi) as angles, users will no longer need edge indices and various angles can be set. Instead, object socket will be needed again and the socket interface will be far different from original grasshopper node interface...

> 7.Thanks

Me too, thank you for insightful suggestions :-).

asahidari commented:
> > 5.Maybe to leave both approaches? object and sverchok.
> > 6.No need to enter edges themselves, we can insert edges' indices. not to fi    lter, less steps for user.
>
> Yes, the node does not need object socket now. Instead, vertices, edges and fa    ces are needed to calculate angles and rotate faces.

Currently I'm changing the node's socket interface.
* Replacing object socket with verts/edges/faces sockets (following the previous suggestions)
* Adding crease-edge indices instead of edge themselves to input valley/mountain edges(following the previous suggestions)
* Adding crease-edge angles instead of a valley angle and a mountain angle. This change enable us to input some kind of angles. For example, water bomb square tessellation in attached image have 2 types of mountain angles (90 and 180 degr    ee.)

Input node will be simple like this.
![rigid_origami_input_indices](https://user-images.githubusercontent.com/64673405/89205883-6666be00-d5f3-11ea-8c49-394ab122a7c5.png)

![Waterbomb_square_tessellation](https://user-images.githubusercontent.com/64673405/89206521-5f8c7b00-d5f4-11ea-9337-bb4d4da0fe1d.png)

Attached current script just in case.
[rigid_origami_folding_mesh_input.py.zip](https://github.com/nortikin/sverchok/files/5017200/rigid_origami_folding_mesh_input.py.zip)

asahidari commented:
Also, I found a limitation (or a bug) only when a paper has multiple vertices inside.

When folding, the node calculates edge angles ('rho' angles in [original paper](https://origami.c.u-tokyo.ac.jp/~tachi/cg/SimulationOfRigidOrigami_tachi_4OSME.pdf)) around vertices inside, and multiply rotation-matrices in counter-clockwise order to get rotation angles. However, folding result will be sometimes changed or cannot be folded when the 'first' edge (in counter-clockwise order) changes. Currently, I'm trying to solve this, but have not been able to find any explanation in the source article.

![rotation_edges_around_verticies_problem_modified](https://user-images.githubusercontent.com/64673405/89208747-ec850380-d5f7-11ea-9378-7180ed6793c9.png)

vicdoval commented:
The interaction with the Pulga Physics would be:
Outputting from the node the force vector that is applied to each point and use it as initial velocity in the Pulga Physics. 
Probably it is too tricky for a regular user and you approach is simpler to understand and with better (and more accurate) results.

![image](https://user-images.githubusercontent.com/10011941/89209708-1f5ed580-d5bf-11ea-9e54-ee8fa7ecaa0a.png)

vicdoval commented:
> So if I can easily write a full node and find where to place it, I want to do that

If you wish you can turn it in into a regular node. I  would recommend you to check a couple of PRs with new nodes as guide
https://github.com/nortikin/sverchok/pull/3263
https://github.com/nortikin/sverchok/pull/3264
https://github.com/nortikin/sverchok/pull/3284

asahidari commented:
@vicdoval , Thanks for the supportive information.

> The interaction with the Pulga Physics would be:
> Outputting from the node the force vector that is applied to each point and use it as initial velocity in the Pulga Physics. 
> Probably it is too tricky for a regular user and you approach is simpler to understand and with better (and more accurate) results.

Initial-velocity vector to each point, I see. It seems the velocity should be given by crease edge rotation. It's not simple, but is cool if the motion is successfully generated. While making the full node (below), I'll try some implementation, but sorry if that could not make successful results.

if the vectors are simply 'verts_out - verts_in' of current node (attached image below), it is easy to implement and it works in some conditions, but it seems not to be the final solution, probably.
![rigid_origami_folding_force_output](https://user-images.githubusercontent.com/64673405/89301320-86a28580-d6a4-11ea-89a7-79e54f6cdd6c.png)

> If you wish you can turn it in into a regular node. I would recommend you to check a couple of PRs with new nodes as guide
> #3263
> #3264
> #3284

Thank you for the help. I can probably make the full node by following the guide. If the current issue would be solved, I'll try it. But please give me some more time (a couple of week, probably) if allowed.

vicdoval commented:
@asahidari  take all the time you need. If anyone is in a hurry they can use the script you provided 😄

asahidari commented:
> @asahidari take all the time you need. If anyone is in a hurry they can use the script you provided 😄

Thank you, @vicdoval , that's right and I feel relieved.

And good news. Fortunately, the problem I mentioned before (below) seems possible to be solved 👍 . 
> Also, I found a limitation (or a bug) only when a paper has multiple vertices inside.

After some more tests, I can start to go on next step (=creating a full node, write documents, etc.).

asahidari committed:
change interfaces to reflect feedbacks

asahidari committed:
Delete rigid origami script file.

asahidari committed:
Add rigid origami full node.

asahidari committed:
Add rigid origami documentation.

asahidari committed:
Avoid null access.

Durman commented on utils/rigid_origami_utils.py:
```diff
@@ -0,0 +1,513 @@
+        
+    # Function to generate inside vertex objects
+    @classmethod
+    def GenerateInsideVertices(cls, obj, crease_lines):
```
```suggestion
    def generate_inside_vertices(cls, obj, crease_lines):
```
According Python convention

asahidari replied:
Thank you. I'll check other function names whether they are defined in the same way.

Durman commented on utils/rigid_origami_utils.py:
```diff
@@ -0,0 +1,513 @@
+        # Create vertex indices inside of the source paper
+        num_verts = len(obj.verts)
+        indices = [v.index for v in obj.bm.verts if not v.is_boundary]
+        InsideVertex.indices = indices
```
```suggestion
        cls.indices = indices
```
It has matter in some cases. I think it would be more correct to use `cls` object.

asahidari replied:
Thank you. I'll check whether any other class objects are used in the same way.

Durman commented on utils/rigid_origami_utils.py:
```diff
@@ -0,0 +1,513 @@
+                                        
+            # Roll the edge order around each vertices
+            for idx, target_ccw in enumerate(target_ccws):
+                if not idx in indices_outermost:
```
```suggestion
                if idx not in indices_outermost:
```
A bit better to read

asahidari replied:
Thank you :). I should have learned more about python before this ...

asahidari committed:
Check list existence in common python way.

Co-authored-by: Durman <sv@soluyanov.ru>

asahidari commented on nodes/modifier_change/rigid_origami.py:
```diff
@@ -0,0 +1,131 @@
+                # Extract crease lines
+                crease_lines = CreaseLines(obj, edge_indices, edge_angles, folding)
+
+                if len(edge_indices) > 0:
```
Thank you for the review :) . I use the way.

asahidari committed:
Check null access in more common way.

Co-authored-by: Durman <sv@soluyanov.ru>

asahidari commented on nodes/modifier_change/rigid_origami.py:
```diff
@@ -0,0 +1,131 @@
+                verts_out.append(verts_o)
+            finally:
+                if obj != None:
+                    obj.free()
```
Thank you for the suggestion. I use the way.

asahidari committed:
Change not idiom a bit better to read.

Co-authored-by: Durman <sv@soluyanov.ru>

asahidari committed:
Correct wrong method definition. Will be deleted if not used.

Co-authored-by: Durman <sv@soluyanov.ru>

asahidari commented on utils/rigid_origami_utils.py:
```diff
@@ -0,0 +1,513 @@
+                if v0 in edge and v1 in edge:
+                    bm_e_index = bm_e.index
+                    break
+        return 
```
Thank you. I'll check whether this method is currently used. If not used, I'll delete this method not to mislead someone later.

asahidari committed:
Change function name in python conventional way.

Co-authored-by: Durman <sv@soluyanov.ru>

asahidari committed:
Use cls object instead of own class name.

Co-authored-by: Durman <sv@soluyanov.ru>

asahidari committed:
Change method name following the change of dependent file.

asahidari committed:
Change not idiom in common python way.

asahidari committed:
Remove non-used function.

asahidari committed:
Use the common style in method names.

asahidari committed:
Add header comments about license.

asahidari committed:
Add origami icons.

asahidari committed:
Improve origami icon files.

asahidari committed:
Fill triangles with white color.

asahidari commented:
Thank you for reviewing, @Durman , I learned a lot.

@Vicdoval , sorry for the waiting, but I think I've complete the origami full node.
Could you please confirm this can be merged ?

These are points I've worried whether I've done correctly.
- I gave up the collaboration with Pulga Physics node this time. I tried with output of initial velocity (=multiplying edge length with radian), but I could not get sound results. I'm sorry, but I attached a script node (not a full node) just in case. [rigid_origami_folding_with_physics.py.zip](https://github.com/nortikin/sverchok/files/5077981/rigid_origami_folding_with_physics.py.zip)
- I couldn't find any style guide for node icons. So following other node icons, I made an icon for origami node with Inkscape.
- Following the Pulga physics, I added this origami node to 'Alpha nodes' . Is this OK?
- When I made this node, I've found, read and followed these guidelines basically.
    - #3263 #3264 #3284 (You gave me in previous comment.)
    - http://nikitron.cc.ua/sverch/html/contributing.html
    - https://github.com/nortikin/sverchok/issues/2301
    - https://github.com/nortikin/sverchok/issues/1834

vicdoval commented on nodes/modifier_change/rigid_origami.py:
```diff
@@ -0,0 +1,130 @@
+    sv_icon = 'SV_ORIGAMI'
+
+    folding_ratio : FloatProperty(
+        name = "Folding ratio",
```
according to PEP8  

> Don't use spaces around the = sign when used to indicate a keyword argument, or when used to indicate a default value for an unannotated function parameter

so: 
```
name="Folding ratio"
description="Folding ratio from 0.0 to 1.0" 
```

and so on

asahidari replied:
Thank you for the suggestion, and sorry that I had missed it. I modified the places which has the same issue at https://github.com/nortikin/sverchok/pull/3416/commits/2d0560d679e47680fad664d99c79d1b42ceb8880 .

vicdoval commented on utils/rigid_origami_utils.py:
```diff
@@ -0,0 +1,506 @@
+    @classmethod
+    # Function to create rotation quaternion
+    # to rotate 'rad' radian around 'n' axis
+    def __rot_quat(cls, rad, n):
+        n = n / np.linalg.norm(n)
+        rad = float(rad)
+        R = np.array([[np.cos(rad)+n[0]*n[0]*(1-np.cos(rad)), \
+                       n[0]*n[1]*(1-np.cos(rad))-n[2]*np.sin(rad), \
+                       n[0]*n[2]*(1-np.cos(rad))+n[1]*np.sin(rad), 0], \
+                      [n[1]*n[0]*(1-np.cos(rad))+n[2]*np.sin(rad), \
+                       np.cos(rad)+n[1]*n[1]*(1-np.cos(rad)), \
+                       n[1]*n[2]*(1-np.cos(rad))-n[0]*np.sin(rad), 0], \
+                      [n[2]*n[0]*(1-np.cos(rad))-n[1]*np.sin(rad), \
+                       n[2]*n[1]*(1-np.cos(rad))+n[0]*np.sin(rad), \
+                       np.cos(rad)+n[2]*n[2]*(1-np.cos(rad)), 0], \
+                       [0, 0, 0, 1]])
+        return R
```
```suggestion
    def __rot_quat(cls, rad, n):
        n = n / np.linalg.norm(n)
        rad = float(rad)
        cos_rad = np.cos(rad)
        sin_rad = np.sin(rad)
        R = np.array([
                       [cos_rad + n[0] * n[0] * (1 - cos_rad), \
                       n[0] * n[1] * (1 - cos_rad) - n[2] * sin_rad , \
                       n[0] * n[2] * (1 - cos_rad) + n[1] * sin_rad , 0], \
                       [n[1] * n[0] * (1 - cos_rad)+ n[2] * sin_rad , \
                       cos_rad + n[1] * n[1] * (1-cos_rad ), \
                       n[1] * n[2] * (1 - cos_rad) - n[0] * sin_rad , 0], \
                       [n[2] * n[0] * (1 - cos_rad) - n[1] * sin_rad , \
                       n[2] * n[1] * (1 - cos_rad) + n[0] * sin_rad , \
                       cos_rad + n[2] * n[2] * (1 - cos_rad), 0], \
                       [0, 0, 0, 1]])
        return R
```

asahidari replied:
Thank you. It had not been easy to read, but your modification made it really simple and readable!

asahidari committed:
Simplify messy quaternion and make it more readable.

Co-authored-by: Victor Doval <10011941+vicdoval@users.noreply.github.com>

vicdoval commented:
![image](https://user-images.githubusercontent.com/10011941/90315298-a0509200-df1a-11ea-8b0e-da05fc232996.png)

@asahidari  Thanks again and congratulations!
Aside of the couple of format suggestions i mention in the code everything seems pretty clear and well documented.

Don't worry about the interaction with the Pulga node, it was a side idea, also Pulga Physics is pretty experimental and it may change in the future that is why it is in the Alpha Category, if your node is stable, has documentation and it is in its final state maybe it would be better to move it to Modifier Make category or Modify Change (it could work in both).

Maybe would be nice to add one or more examples (exported jsons) to the examples folder in the category you find more suitable for them :)

asahidari committed:
Remove spaces around equal signs used to indicate keyword arguments.

asahidari committed:
Move the rigid origami node from alpha to final list.

asahidari committed:
Add an origami windmill example in document.

asahidari commented:
> @asahidari Thanks again and congratulations!
> Aside of the couple of format suggestions i mention in the code everything seems pretty clear and well documented.

Thank you @vicdoval, I appreciated your review. I was relieved to hear that. (And the image you attached is photorealistic and very cool!  😄)

> Don't worry about the interaction with the Pulga node, it was a side idea, also Pulga Physics is pretty experimental and it may change in the future that is why it is in the Alpha Category, if your node is stable, has documentation and it is in its final state maybe it would be better to move it to Modifier Make category or Modify Change (it could work in both).

OK, I agree with you about the physics node. And I just moved it from Alpha Category to Modify Change at https://github.com/nortikin/sverchok/pull/3416/commits/d120e5c110f356380eee8f3dd938d7465417b3c2 .
 
> Maybe would be nice to add one or more examples (exported jsons) to the examples folder in the category you find more suitable for them :)

I agree it is good for users to import examples from json files. However, source objects cannot be exported and the edge indices might be different between users. So, I displayed source objects and the edge indices in example images instead. Is it OK without exported json files in this case? Anyway, I added another example at https://github.com/nortikin/sverchok/pull/3416/commits/149e179485407959a4717a2c345e444b607fb156 .

asahidari committed:
Add water bomb tessellation example in document.

asahidari committed:
Add origami json examples.

asahidari commented:
> > Maybe would be nice to add one or more examples (exported jsons) to the examples folder in the category you find more suitable for them :)

> I agree it is good for users to import examples from json files. However, currently source objects are not exported together and the edge indices might be different between users. So, I displayed source objects and the edge indices in example images instead. Is it OK without exported json files in this case?

Sorry, I did not know Text In/Out node can dump/load object mesh datas. (These are very useful! I found that when I try other json_examples. And even without these nodes, simple data can be made with vector/edge indices list/polygon indices list.)

I added two examples at https://github.com/nortikin/sverchok/pull/3416/commits/6eea52089e629cf4c835b1e0b5fe1d106b343401 . Could you check these once more?

asahidari committed:
Add namako origami tessellation example.

vicdoval commented:
@asahidari  Im glad you found the way to add some examples, they are awesome.

The PR is almost ready 😄 , the only thing is that there is a duplicated example:
Water bomb origami tessellation and Water bomb origami tessellation square

asahidari committed:
Remove an old origami tessellation example.

asahidari commented:
> @asahidari Im glad you found the way to add some examples, they are awesome.
> 
> The PR is almost ready 😄 , the only thing is that there is a duplicated example:
> Water bomb origami tessellation and Water bomb origami tessellation square

Thanks for the pointing out, and I'm sorry. Adding examples is backfired... All I wanted to do has been completed, and will do nothing if there are no need. Really, sorry 😰.

vicdoval commented:
@asahidari  Thanks and congratulations for your first full featured node in Sverchok!!

nortikin commented:
lovely

asahidari commented:
This experience had been very exciting!

@vicdoval , thank you for many supports and advices!

@nortikin , thank you for many suggestions and approving this PR!

@enzyme69 , thank you for trying this node and suggesting a lot!

@Durman , thank you for reviewing and many advices!

asahidari commented:
@nortikin 
I have to apologize to you. While I was involved in this PR, I have missed all mails from GitHub (I noticed those mails some days after this PR was merged!). Some were important and I probably missed some opportunities. Despite my this rude attitude, however, you have involved in this PR. I sincerely appreciated, thanks.

mavizasyon commented:
How can I do this flasher origami? I have zero knowledge about sverchok nodes and tutorials are seem to outdated. Red mountain, blue valley
![Eskiz1](https://github.com/user-attachments/assets/5cc8a84f-bd98-467e-880a-65c45a08c7a1)
[flasher.zip](https://github.com/user-attachments/files/19651485/flasher.zip)

asahidari commented:
@mavizasyon

Would you try to see the attached file? Or follow the steps bellow (, images below might help you understand):

1. Fill the center hole in advance to be used as a fixed face for 'Rigid origami' node parameter.
2. Use 'Viewer index+' node to see all edge indices of your polygon.
3. Create list of mountain-edge and valley-edge indices with the final angles of all edges.
4. Feed the parameters for 'Rigid origami' node with connecting nodes and 'Viewer Draw' node to see the result.
5. Change 'Folding ratio' parameter from 0.0 to 1.0 to see the folding step.

I tried it and saw some results. When 'Folding ratio' in 'Rigid origami' node is from 0.0 to 0.3, the result seems well. But when the ratio is over 0.4, the output were not so good and some adjustments (changing final angles, etc.) might be needed.

I hope this comment will help you.

![Screen Shot 2025-04-09 at 21 10 50](https://github.com/user-attachments/assets/74e7ce0f-d272-4e0a-b46f-1557c02dcadc)
![Screen Shot 2025-04-09 at 21 20 40](https://github.com/user-attachments/assets/fc5659b5-b4ce-4232-8ca7-c0d0dc484fd2)
![Screen Shot 2025-04-09 at 21 26 20](https://github.com/user-attachments/assets/220a1891-2051-4cbc-8485-a82614e1471b)
[flasher_modified.blend.zip](https://github.com/user-attachments/files/19667886/flasher_modified.blend.zip)

mavizasyon commented:
@asahidari 

could you help me to create shrinking cylinder origami? I have tried but it freezes a lot and couldn't make it work perfectly, sometimes nothing happens sometimes gives errors. It is hard look and write edge indicies. Is there select and create list option? And another question is how can I achieve self collison like origami simulator.

<img width="552" height="641" alt="image" src="https://github.com/user-attachments/assets/ab726552-014d-4ea7-a414-7a931c01ced8" />

Files;

https://drive.google.com/file/d/1E1F5y1pzQrXib7RH0HmVs2QX0z9YHC6v/view?usp=sharing
https://drive.google.com/file/d/1C2vfBMPwy3D3Xxl9FGQM7yfOxGgeqxwz/view?usp=sharing

asahidari commented:
@mavizasyon

I'm so sorry to be late for replying. And unfortunately, I do not have any environments to check your cylinder now. So, I can only speculate about your issue, but I try to answer.

* Try to make shorter size cylinders at first. The cylinder, in the image you posted, seems to have already eight sections in vertical direction. If the model has more complicated structure, the more difficult to find the problem. So it seems better to start with simple models to find the solution.

* Check the surface direction (back or front) of each faces of your model. If some faces have different direction, they might cause problems.

If these advice will not work well, I cannot find any solutions right now. And the problem might be caused by the limitation of this node(, because this node is for "rigid" origami, not for lenient, pliable origami).

For the other question you posted: unfortunately, this node is currently not capable of self collision detection. This node is not based on Origami Simulator, because the simulator uses GPU.
